Tests/BuildTests/IncrementalBuildTests.swift

```diff
@@ -15,7 +15,130 @@ import Basic
 import Utility
 import func libc.sleep
 
+/// returns "line# \\(#line)"
+func line(lineNum: Int = #line) -> String {
+    return "line# \(lineNum)"
+}
+
+/// Some common error messages.
+enum IncBuildErrorMessages : String, CustomStringConvertible, Hashable {
+    static func == (lhs: IncBuildErrorMessages, rhs: IncBuildErrorMessages) -> Bool { return lhs.rawValue == rhs.rawValue }
+    var description: String { return self.rawValue }
+    var hashValue: Int { return self.rawValue.hashValue }
+    case fullBuildSameAsIncrementalBuild = "A full build log was the same as an incremental build log"
+    case unexpectedSimilar = "Unexpectedly found two similar build logs"
+    case unexpectedDissimilar = "Unexpectedly found two dissimilar build logs"
+    case unexpectedBuild = "Unexpected build"
+    case unexpectedNullBuild = "Unexpected null build"
+    case unexpectedBuildFailure = "Unexpected build failure"
+    case unexpectedBuildSuccess = "Unexpected build success"
+    case builtTooMany = "The build log contains too many entries"
+    case builtProductsDiffer = "The ./build/debug/ directories are materially different"
+}
+
+/// Some common build names
+enum Builds : String, CustomStringConvertible, Hashable {
+    static func == (lhs: Builds, rhs: Builds) -> Bool { return lhs.rawValue == rhs.rawValue }
+    var description: String { return self.rawValue }
+    var hashValue: Int { return self.rawValue.hashValue }
+    case initial = "Initial Full Build"
+    case withError = "Build with an Error"
+    case withErrorFixed = "Build with the Error Fixed"
+    case null = "Null Build"
+    case withDependency = "Build with a dependency"
+    case withoutDependency = "Build without a dependency"
+    case withTargetDependency = "Build with a target dependency"
+    case withoutTargetDependency = "Build without a target dependency"
+    case withFile = "Build with a file"
+    case withoutFile = "Build without a file"
+}
+
+/// Set to true to have every test do a "redudant" build that isn't expected
+/// to do anything. We have a test that explicitly checks regardless of this
+/// value, so it should be safe to leave it as false
+var alwaysDoRedundancyCheck = false
+
+/// returns an array of keys whose value is nil. Pass an array of keys to
+/// `allowableFailures` if it's not an error for something to be nil.
+@discardableResult
+func checkForNils(_ logs: BuildLog, allowableFailures: [Builds] = []) -> [Builds] {
+    return logs.keys.filter {
+        (logs[$0] == nil && !($0 == Builds.null && !alwaysDoRedundancyCheck)) || (logs[$0]!.logWasNil && !allowableFailures.contains($0))
+    }
+}
+
+typealias BuildLog = [Builds : BuildResult]
 
+// Seems like this is too tied to the build's log format
```

Comment:
Agreed, we should eventually have a mode for llbuild that just dumps a streaming JSON log so it can be parsed mechanically.